From 45c30c790c7921a25a5c615eef614bb40d5096a4 Mon Sep 17 00:00:00 2001
From: Shixiang Zheng <shixiang.zheng@rock-chips.com>
Date: Tue, 30 Oct 2018 17:18:34 +0800
Subject: [PATCH] video/rockchip: rga2: fix rgaColorFill bug and memory check
 bug

front 2736, back 68480 dsize 65536 size1824
RGA2 Get MMu mem failed
rga2_reg_init, [770] set mmu info error
rga: init reg fail

Change-Id: I6feaf1f5c176b1eb2ec1122026fd7e1aa96dfba1
Signed-off-by: Shixiang Zheng <shixiang.zheng@rock-chips.com>
---
 drivers/video/rockchip/rga2/rga2_drv.c      | 10 +++++-----
 drivers/video/rockchip/rga2/rga2_mmu_info.c |  3 ++-
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/video/rockchip/rga2/rga2_drv.c b/drivers/video/rockchip/rga2/rga2_drv.c
index e8580a381352..0b32f00bb961 100644
--- a/drivers/video/rockchip/rga2/rga2_drv.c
+++ b/drivers/video/rockchip/rga2/rga2_drv.c
@@ -404,15 +404,15 @@ static int rga2_memory_check(void *vaddr, u32 w, u32 h, u32 format, int fd)
 		kfree(one_line);
 		return -1;
 	}
-	temp_data = w * (h - 1) * bits / 3;
+	temp_data = w * (h - 1) * bits >> 3;
 	if (fd > 0) {
 		DBG("vaddr is%p, bits is %d, fd check\n", vaddr, bits);
-		memcpy(one_line, (char *)vaddr + temp_data, w * bits / 3);
+		memcpy(one_line, (char *)vaddr + temp_data, w * bits >> 3);
 		DBG("fd check ok\n");
 	} else {
 		DBG("vir addr memory check.\n");
 		memcpy((void *)((char *)vaddr + temp_data), one_line,
-		       w * bits / 3);
+		       w * bits >> 3);
 		DBG("vir addr check ok.\n");
 	}
 	kfree(one_line);
@@ -1181,7 +1181,7 @@ static int rga2_convert_dma_buf(struct rga2_req *req)
 	if (RGA2_CHECK_MODE) {
 		vaddr = ion_map_kernel(rga2_drvdata->ion_client, hdl);
 		if (vaddr)
-			rga2_memory_check(vaddr, req->src.vir_h, req->src.vir_w,
+			rga2_memory_check(vaddr, req->src.vir_w, req->src.vir_h,
 					  req->src.format, req->src.yrgb_addr);
 		ion_unmap_kernel(rga2_drvdata->ion_client, hdl);
 	}
@@ -1223,7 +1223,7 @@ static int rga2_convert_dma_buf(struct rga2_req *req)
 	if (RGA2_CHECK_MODE) {
 		vaddr = ion_map_kernel(rga2_drvdata->ion_client, hdl);
 		if (vaddr)
-			rga2_memory_check(vaddr, req->dst.vir_h, req->dst.vir_w,
+			rga2_memory_check(vaddr, req->dst.vir_w, req->dst.vir_h,
 					  req->dst.format, req->dst.yrgb_addr);
 		ion_unmap_kernel(rga2_drvdata->ion_client, hdl);
 	}
diff --git a/drivers/video/rockchip/rga2/rga2_mmu_info.c b/drivers/video/rockchip/rga2/rga2_mmu_info.c
index 9570292e32b1..61673e19659c 100644
--- a/drivers/video/rockchip/rga2/rga2_mmu_info.c
+++ b/drivers/video/rockchip/rga2/rga2_mmu_info.c
@@ -833,7 +833,8 @@ static int rga2_mmu_info_color_fill_mode(struct rga2_reg *reg, struct rga2_req *
         /* flush data to DDR */
         rga_dma_flush_range(MMU_Base, (MMU_Base + AllSize + 1));
         rga2_mmu_buf_get(&rga2_mmu_buf, AllSize);
-
+	reg->MMU_len = AllSize;
+
         return 0;
     }
     while(0);
-- 
2.35.3

